datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -87,6 +138,41 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
 
     Validate.isTrue(totalTaskCount > 0, String.format("No tasks found for datastream group %s", dgName));
 
+    int minTasks = getMinTasks(datastreamPartitions.getDatastreamGroup());
+    if (getEnableElasticTaskAssignment(minTasks) && assignedPartitions.isEmpty()) {
+      // The partitions have not been assigned to any tasks yet and elastic task assignment has been enabled for this
+      // datastream. Assess the number of tasks needed based on partitionsPerTask and the fullness threshold. If
+      // the number of tasks needed is smaller than the number of tasks found, throw a DatastreamRuntimeException
+      // so that LEADER_DO_ASSIGNMENT and LEADER_PARTITION_ASSIGNMENT can be retried with an updated number of tasks.
+      int partitionsPerTask = resolveConfigWithMetadata(datastreamPartitions.getDatastreamGroup(),
+          CFG_PARTITIONS_PER_TASK, _partitionsPerTask);
+      int partitionFullnessFactorPct = resolveConfigWithMetadata(datastreamPartitions.getDatastreamGroup(),
+          CFG_PARTITION_FULLNESS_THRESHOLD_PCT, _partitionFullnessFactorPct);
+      LOG.info("Calculating number of tasks needed based on partitions per task: calculated->{}:config->{}, "
+              + "fullness percentage: calculated->{}:config->{}", partitionsPerTask, _partitionsPerTask,
+          partitionFullnessFactorPct, _partitionFullnessFactorPct);
+      int allowedPartitionsPerTask = (partitionsPerTask * partitionFullnessFactorPct) / 100;
+      int totalPartitions = datastreamPartitions.getPartitions().size();
+      int numTasksNeeded = (totalPartitions / allowedPartitionsPerTask)
+          + (((totalPartitions % allowedPartitionsPerTask) == 0) ? 0 : 1);
+      int maxTasks = resolveConfigWithMetadata(datastreamPartitions.getDatastreamGroup(), CFG_MAX_TASKS, 0);
+      if ((maxTasks > 0) && (numTasksNeeded > maxTasks)) {
```

Comment:
The comment is confusing since this if is already inside another if check which checks for elastic task assignment to be enabled.

Reply:
moved this explanation to the top of the file